File: Android/PitchPerfect/src/depollsoft/pitchperfect/LoginPrompt.java
package depollsoft.pitchperfect;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;

import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.login.LoginResult;
import com.facebook.login.widget.LoginButton;
import com.parse.ParseUser;
import com.parse.boltsinternal.Continuation;
import com.parse.facebook.ParseFacebookUtils;

import bolts.Capture;

public class LoginPrompt {
  public static final CallbackManager FACEBOOK_CALLBACK_MANAGER = CallbackManager.Factory.create();

  private static void completeLogin(boolean isNew) {
    if (!isNew) {
      SettingsModel.restoreUser();
      SongsModel.get().refreshFromParse();
    } else {
      SettingsModel.refreshUser();
      SongsModel.get().saveAllToParse(true);
    }
  }

  public static Dialog buildDialog(final Context context, boolean isHoomiLogout) {
    View view = LayoutInflater.from(context).inflate(R.layout.loginpromptview, null);

    final Capture<AlertDialog> dialog = new Capture<>(null);

    final LoginButton fbLoginButton = (LoginButton) view.findViewById(R.id.fb_login_button);
    // Callback registration
    fbLoginButton.registerCallback(FACEBOOK_CALLBACK_MANAGER, new FacebookCallback<LoginResult>() {
      @Override
      public void onSuccess(final LoginResult loginResult) {
        final ProgressDialog progressDialog = new ProgressDialog(context);
        progressDialog.setMessage("Please wait...");
        progressDialog.show();
        ParseFacebookUtils.logInInBackground(loginResult.getAccessToken())
            .onSuccess((Continuation<ParseUser, Void>) task -> {
              completeLogin(task.getResult().isNew());
              if (dialog.get().isShowing()) {
                dialog.get().dismiss();
              }
              return null;
            }).continueWith((Continuation<Void, Void>) task -> {
              if (progressDialog.isShowing()) {
                progressDialog.dismiss();
              }
              return null;
            });
      }

      @Override
      public void onCancel() {
        // App code
      }

      @Override
      public void onError(FacebookException exception) {
        // App code
      }
    });

    TextView explanationText = (TextView) view.findViewById(R.id.explanationTextView);
    String explanation = context.getResources().getString(isHoomiLogout ?
        R.string.HoomiLoginExplanation : R.string.LoginExplanation);
    explanationText.setText(Html.fromHtml(explanation));

    dialog.set(new AlertDialog.Builder(context)
        .setView(view)
        .setNeutralButton(R.string.SkipLogin, new DialogInterface.OnClickListener() {
              @Override
              public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
              }
            }
        )
        .setTitle(R.string.LoginTitle)
        .create());
    return dialog.get();
  }
}

File: Android/PitchPerfect/src/depollsoft/pitchperfect/LoginPrompt.kt
package depollsoft.pitchperfect

import android.app.Activity.RESULT_OK
import android.app.Dialog
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import androidx.activity.ComponentActivity
import androidx.appcompat.app.AlertDialog
import bolts.Capture
import com.firebase.ui.auth.AuthUI
import com.firebase.ui.auth.FirebaseAuthUIActivityResultContract
import com.firebase.ui.auth.data.model.FirebaseAuthUIAuthenticationResult

object LoginPrompt {
    fun buildDialog(activity: ComponentActivity, isHoomiLogout: Boolean): Dialog {
        val dialog = Capture<AlertDialog?>(null)
        val launcher = activity.registerForActivityResult(
            FirebaseAuthUIActivityResultContract()
        ) { result ->
            if (result.resultCode == RESULT_OK) {
                dialog.get()?.dismiss()
                if (!result.idpResponse!!.isNewUser) {
                    SettingsModel.attachToFirestore()
                    SongsModel.get().attachToFirestore(false)
                } else {
                    SettingsModel.attachToFirestore()
                    SongsModel.get().attachToFirestore(true)
                }
            }
        }
        val view = LayoutInflater.from(activity).inflate(R.layout.loginpromptview, null)
        val loginButton = view.findViewById<View>(R.id.login_button)
        loginButton.setOnClickListener {
            val logInIntent = AuthUI.getInstance().createSignInIntentBuilder()
                .setIsSmartLockEnabled(false)
                .setAvailableProviders(
                    listOf(
                        AuthUI.IdpConfig.EmailBuilder()
                            .setRequireName(false)
                            .setAllowNewAccounts(true)
                            .build(),
                        AuthUI.IdpConfig.GoogleBuilder().build(),
                        AuthUI.IdpConfig.FacebookBuilder().build(),
                    )
                )
                .setTheme(R.style.AppTheme)
                .build()
            launcher.launch(logInIntent)
        }
        // Callback registration

        val explanationText = view.findViewById<View>(R.id.explanationTextView) as TextView
        val explanation =
            activity.resources.getString(if (isHoomiLogout) R.string.HoomiLoginExplanation else R.string.LoginExplanation)
        explanationText.text = if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            Html.fromHtml(explanation, Html.FROM_HTML_MODE_LEGACY)
        } else {
            @Suppress("DEPRECATION")
            Html.fromHtml(explanation)
        }
        dialog.set(AlertDialog.Builder(activity)
            .setView(view)
            .setNeutralButton(
                R.string.SkipLogin
            ) { dialog, which -> dialog.dismiss() }
            .setTitle(R.string.LoginTitle)
            .create())
        return dialog.get()!!
    }
}